src/client/auth/oidc.rs

```diff
@@ -168,68 +172,79 @@ pub struct IdpServerResponse {
     pub refresh_token: Option<String>,
 }
 
-pub(crate) fn build_speculative_client_first(credential: &Credential) -> Command {
-    self::build_client_first(credential, None)
-}
-
-/// Constructs the first client message in the OIDC handshake for speculative authentication
-pub(crate) fn build_client_first(
-    credential: &Credential,
+fn make_spec_auth_command(
+    source: String,
+    payload: Vec<u8>,
     server_api: Option<&ServerApi>,
 ) -> Command {
-    let mut auth_command_doc = doc! {
-        "authenticate": 1,
+    let body = doc! {
+        "saslStart": 1,
         "mechanism": MONGODB_OIDC_STR,
+        "payload": Binary { subtype: BinarySubtype::Generic, bytes: payload },
+        "db": "$external",
     };
 
-    if credential.oidc_callback.is_none() {
-        auth_command_doc.insert("jwt", "");
-    } else if let Some(access_token) = get_access_token(credential) {
-        auth_command_doc.insert("jwt", access_token);
-    }
-
-    let mut command = Command::new("authenticate", "$external", auth_command_doc);
+    let mut command = Command::new("saslStart", source, body);
     if let Some(server_api) = server_api {
         command.set_server_api(server_api);
     }
-
     command
 }
 
-fn get_access_token(credential: &Credential) -> Option<String> {
-    credential
-        .oidc_callback
-        .as_ref()
-        .unwrap()
-        .cache
-        .read()
-        .unwrap()
-        .access_token
-        .clone()
+pub(crate) async fn build_speculative_client_first(credential: &Credential) -> Option<Command> {
+    self::build_client_first(credential, None).await
 }
 
-fn get_refresh_token_and_idp_info(
+/// Constructs the first client message in the OIDC handshake for speculative authentication
+pub(crate) async fn build_client_first(
     credential: &Credential,
```

Comment:
This had to go `async` because we needed to use a tokio `Mutex` in order to properly handle the criticality necessary to make the tests pass. The change from `RwLock` to `Mutex` is due to the overall criticality necessary. There really is no advantage to a `RwLock` in this algorithm